### ISSUE TITLE
Hook to transform ids...

### ISSUE THREAD

**Issue by joeyhub** [rollup/rollup#3114]
# 😥

Although people do things in node_modules every which way, there's a common tendency to point main to a prebuild resulting from any one of a million and one notions of how something should be built for distribution, often attempting to cater to everything.

The distribution versions however are often needless where a combination of rollup, public and private plugins are used. The obvious example might be that something might export a dist in the form of CJS compiled from ESM which would be unnecessary for use with rollup.

Most problematic cases are where dist versions contain many polyfills and other things that are not needed or handler elsewhere. This as the least adds bloat and at the worst causes breakages.

An example is a package I'm using where it issues a dist with all the works to make it browser compatible. Both module and main point to prebuilds. It includes however the source as ESM and TS. These are easily handled by my babel pass on the final bundle (default builds are not in compatibility mode which makes things like debugging a lot easier).

In another case, I remap names to either _fore or _aft based on whether the build is for fore or aft. Includes exclude the suffix and a plugin then adds the suffix as is appropriate if a file with the appropriate suffix exists.

I want things to work without compilation on _aft where necessary so I also would like to have a file without the suffixes which would serve as a proxy (can even be a symlink) to allow code to execute using ESM by default without the need to rollup.

# 😱

The resolve hook exists only to find non-existent dependencies which means that once it finds a dependency there's no further opportunity to transform the path, at least explicitly.

Some workarounds exist but they're not ideal...

* Point to non-existent things on purpose. Create horrific kludges and wont work with existing libs. Is also potentially single pass (not sure what happens when resolve returns something that is a string but doesn't exist).
* Use load. Problematic as it only runs once and may cause the file watch, dependencies, etc to be wrong.

Resolve can work on a first pass but not on a second pass without reimplementing plugins such as resolve.

# 😍

What we need is a translateId hook. It would take the same input. When null is returned it would be ignored. When a string is returned it would accept that as the new id, repeat resolve is necessary then continue running the other hooks.

There is some debate on whether running resolve again is appropriate though I'd say probably it is unless the id was resolved already.

**Comment by joeyhub** [comment]
In the mean time I'm going to have to make a plugin that proxies the other plugins that are resolving.

**Comment by shellscape** [comment]
@joeyhub thanks for the very extensive issue write-up. Please _do_ use one of the issue templates next time though.

**Comment by lukastaegert** [comment]
@joeyhub I agree with @TrySound here. It is not the solution you propose but to me it sounds like the [`this.resolve`](https://rollupjs.org/guide/en/#thisresolvesource-string-importer-string-options-skipself-boolean--promiseid-string-external-boolean--null) plugin context function would solve your issue. Note also that there is an option to make sure a plugin does not trigger its own resolver when using the context function.

**Comment by joeyhub** [comment]
Side note: The annoying thing with the template is that its preamble is practically a signature with a lot of spam /  junk / stuff people usually have learnt to simply filter out. It'll take up the entire input then some before it finally gets to the point of issue layout which is a simple three lines of sections. My default I'm going to see immediately it's noise blocking the input box which is for the user to input into then ctrl + a and delete also taking out the bottom bit which is actually useful. Instead what you really want is github to properly support templates and for templates to be well thought out (IE, some template approaches can be too inflexible).

Wrapping resolve and proxying plugins works. The first past of resolveId runs no matter what (whether something is found or not) from what I see even on sources that should not need to be resolved. I'm not sure this is documented.

The resolve plugin has an alias feature / option for simple cases (but not named alias) which can solve a lot of use cases, though I don't believe it can do things such as give the option to operate based om importer or resolved name (need to check the last one), it appears to purely by importee.

**Comment by lukastaegert** [comment]
> The resolve plugin has an alias feature / option

Not sure we are talking about the same thing here. The [`this.resolve`](https://rollupjs.org/guide/en/#thisresolvesource-string-importer-string-options-skipself-boolean--promiseid-string-external-boolean--null) context function resolves based on imported file (called `source` in the documentation) and importer together, exactly like `resolveId`. As `source` you can pass any kind of id you like—unresolved, resolved once, resolved ten times…

**Comment by joeyhub** [comment]
Something akin to that. Similar to the alias plugin but with more flexibility. I'm sure I've seen somewhere people giving aliases with the resolve plugin as well somehow.

I believe the problem is probably solved or at least I seem to be able to write a plugin that lets me do whatever I want given that `resolveId` is called all the time for first seen items whether it managed to resolve it or not. Basically, the built in resolve is put at the end of the list rather than at start.

Maybe it's more a DOC issue around SOC since "resolve" and "alias" ("transform") settle two different concerns but the docs aren't clear that it can be used for both through deferring its own resolve as a fallback.
